cli-plugins/manager/hooks.go

```diff
@@ -0,0 +1,180 @@
+type HookPluginData struct {
+	RootCmd string
+}
+
+// HookMessage represents a plugin hook response. Plugins
+// declaring support for CLI hooks need to print a json
+// representation of this type when their hook subcommand
+// is invoked.
+type HookMessage struct {
+	// TODO: should we have an enum to typify the kind of hook
+	// (such as `Next Steps`, `Warning`, etc.)?
+	Template string
+}
+
+// TemplateReplaceSubcommandName returns a hook template string
+// that will be replaced by the CLI subcommand being executed
+//
+// Example:
+//
+// "you ran the subcommand: " + TemplateReplaceSubcommandName()
+//
+// when being executed after the command:
+// `docker run --name "my-container" alpine`
+// will result in the message:
+// `you ran the subcommand: run`
+func TemplateReplaceSubcommandName() string {
+	return hookTemplateCommandName
+}
+
+// TemplateReplaceFlagValue returns a hook template string
+// that will be replaced by the flags value.
+//
+// Example:
+//
+// "you ran a container named: " + TemplateReplaceFlagValue("name")
+//
+// when being executed after the command:
+// `docker run --name "my-container" alpine`
+// will result in the message:
+// `you ran a container named: my-container`
+func TemplateReplaceFlagValue(flag string) string {
+	return fmt.Sprintf(hookTemplateFlagValue, flag)
+}
+
+// TemplateReplaceArg takes an index i and returns a hook
+// template string that the CLI will replace the template with
+// the ith argument, after processing the passed flags.
+//
+// Example:
+//
+// "run this image with `docker run " + TemplateReplaceArg(0) + "`"
+//
+// when being executed after the command:
+// `docker pull alpine`
+// will result in the message:
+// "Run this image with `docker run alpine`"
+func TemplateReplaceArg(i int) string {
+	return fmt.Sprintf(hookTemplateArg, strconv.Itoa(i))
+}
+
+// RunPluginHooks calls the hook subcommand for all present
+// CLI plugins that declare support for hooks in their metadata
+// and parses/prints their responses
+func RunPluginHooks(dockerCli command.Cli, rootCmd, subCommand *cobra.Command) error {
+	plugins, err := ListPlugins(dockerCli, rootCmd)
+	if err != nil {
+		return err
+	}
+	subCmdName := subCommand.Name()
+
+	hookMessages := make([]string, 0, len(plugins))
+	for _, p := range plugins {
+		p := p
+		if !p.supportsHookCommand(subCmdName) {
+			continue
+		}
+
+		hookReturn, err := p.RunHook(subCmdName)
+		// Skip misbehaving plugins, but don't halt execution
+		if err != nil {
+			continue
+		}
+		hookMessages = append(hookMessages, parseHooks(hookReturn.Template, subCommand))
+	}
+
+	printHookMessages(dockerCli.Err(), hookMessages)
+	return nil
+}
+
+const (
+	hookTemplateCommandName = "{docker-cli-plugin-hook-cmd-name}"
+	hookTemplateFlagValue   = "{docker-cli-plugin-hook-flag-%s}"
+	hookTemplateArg         = "{docker-cli-plugin-hook-arg-%s}"
+)
+
+func parseHooks(template string, cmd *cobra.Command) string {
+	ret := template
+	for {
```

Comment:
nit: maybe we could leave a short comment here about why we run `parseFlagValue` and `parseArg` in a loop that we exit when we don't find any more matches, since the reason seems to be how the regex lib itself works inside of those funcs. WDYT?

Reply:
Hmmmm, I'm generally of the opinion that comments explaining what code is doing (not why it is doing something) are a yellow flag: if we feel the need to do it, we should definitely consider whether the code couldn't be refactored to be clearer. I just added some hooks tests, so I'll make a run at a refactor here and if it's still unclear I'll add comments/change some stuff around.

Reply:
Same as below, did a larger refactor here (now under `cli-plugins/hooks/template.go`, PTAL and let me know if the current implementation is clearer.